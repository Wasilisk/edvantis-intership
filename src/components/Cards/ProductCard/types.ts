import {ProductType} from "../../../models";
import {ButtonVariants} from "../../common/Button/types";

export type ProductCardProps = Omit<ProductType, "additional_info"> & { buttonVariant: ButtonVariants }