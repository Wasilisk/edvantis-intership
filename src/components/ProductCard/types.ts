import {ProductType} from "../../models";
import {ButtonVariants} from "../common/Button/types";

export type ProductCardProps = {
    content: ProductType,
    buttonVariant: ButtonVariants
}