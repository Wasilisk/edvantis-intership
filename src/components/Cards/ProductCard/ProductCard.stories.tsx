import {Story} from "@storybook/react";
import React from "react";
import ProductCard from "./index";
import {ProductCardProps} from "./types";

const mockProductCard = {
    id: 1,
    search_name: "lord-armchair-miedel-home",
    name: "Lord armchair, Miedel Home",
    price: 120.5,
    image: {
        url: "https://e-c.storage.googleapis.com/res/5f731c835a5a2667cb27290e/480"
    }
}

export default {
    title: 'Application/Cards/ProductCard',
    component: ProductCard,
    decorators: [
        (Story: Story) => (
            <div style={{width: "30%"}}>
                <Story/>
            </div>
        )
    ]
};

const Template: Story<ProductCardProps> = (props) =>
    <ProductCard
        {...props}
        buttonVariant={props.buttonVariant}
    />;

export const WithOutlinedButton = Template.bind({});
WithOutlinedButton.args = {
    ...mockProductCard,
    buttonVariant: "outlined"
};

export const WithFilledButton = Template.bind({});
WithFilledButton.args = {
    ...mockProductCard,
    buttonVariant: "filled"
};