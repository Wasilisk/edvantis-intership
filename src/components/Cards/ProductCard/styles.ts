import styled from "styled-components";

export const ProductCardStyles = styled.div`
  display: flex;
  justify-content: flex-start;
  align-items: center;
  flex-direction: column;
  
  .card-text {
    text-align: center;
    margin-top: 24px;
    
    a {
      color: #333333;
      text-decoration: none;
      overflow-wrap: break-word;
      word-wrap: break-word;
      font-size: 20px;
      line-height: 1.33;
      font-weight: 500;
      letter-spacing: .05em;
      text-transform: none;
    }
    
    p {
      font-weight: 700;
      font-size: 20px;
      white-space: nowrap;
      margin-top: 16px;
      margin-bottom: 24px;
    }
  }
  
  .card-image {
    position: relative;
    width: 100%;
    height: 0;
    padding-bottom: 100%;
    float: left;
    
    
    img {
      position: absolute;
      height: 100%;
      left: 50%;
      transform: translate(-50%, 0);
      object-fit: contain;
    }
    
  }
`