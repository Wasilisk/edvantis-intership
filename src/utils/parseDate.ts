const month = ["January","February","March","April","May","June","July","August","September","October","November","December"];

export const parseDate = (timestamp: number) => {
    const date = new Date(timestamp);

    return `${month[date.getMonth()]} ${('0' + date.getDate()).slice(-2)}, ${date.getFullYear()}`;
}