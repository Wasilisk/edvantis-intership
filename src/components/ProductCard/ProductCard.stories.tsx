import {Story} from "@storybook/react";
import React from "react";
import {BrowserRouter} from "react-router-dom";
import ProductCard from "./index";
import {ProductCardProps} from "./types";

const mockProductCard = {
    id: 1,
    search_name: "lord-armchair-miedel-home",
    name: "Lord armchair, Miedel Home",
    price: 120.5,
    image: {
        url: "https://e-c.storage.googleapis.com/res/5f731c835a5a2667cb27290e/480"
    }
}

export default {
    title: 'Application/ProductCard',
    component: ProductCard,
    decorators: [
        (Story: Story) => (
            <div style={{ width: "300px" }}>
                <Story />
            </div>
        ),
    ],
};

const Template: Story<ProductCardProps> = (props) => <BrowserRouter>
    <ProductCard
    content={props.content}
    buttonVariant={props.buttonVariant}
/>
</BrowserRouter>;

export const WithOutlinedButton = Template.bind({});
WithOutlinedButton.args = {
    content: mockProductCard,
    buttonVariant: "outlined"
};

export const WithFilledButton = Template.bind({});
WithFilledButton.args = {
    content: mockProductCard,
    buttonVariant: "filled"
};