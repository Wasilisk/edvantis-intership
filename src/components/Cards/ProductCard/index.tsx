import React from 'react';
import { ProductCardStyles } from './styles';
import {ProductCardProps} from "./types";
import Button from "../../common/Button";
import {Link} from "react-router-dom";

const ProductCard = ({search_name, image, name, price, buttonVariant, status}: ProductCardProps) => {

    const clickHandler = () => console.log(search_name); //ToDo add product card logic

    return (
        <ProductCardStyles data-testid="product-card">
            <Link className="card-image" to={`shop/${search_name}`}>
                <img src={image?.url} alt={name}/>
                {
                    status === "out of stock" && <div className="out-of-stock">Out of stock</div>
                }
            </Link>
            <div className="card-text">
                <Link to={`shop/${search_name}`}>{name}</Link>
                <p className="highlight">
                    {price.toFixed(2)}USD
                </p>
            </div>
            <Button
                variant={buttonVariant}
                size="small"
                label="Order Now"
                onClick={clickHandler}
                disabled={status === "out of stock"}
            />
        </ProductCardStyles>
    );
};

export default ProductCard;