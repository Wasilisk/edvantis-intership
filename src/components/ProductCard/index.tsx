import React from 'react';
import { ProductCardStyles } from './styles';
import {ProductCardProps} from "./types";
import Button from "../common/Button";
import {Link} from "react-router-dom";

const ProductCard = ({content, buttonVariant}: ProductCardProps) => {

    const clickHandler = () => console.log(content.search_name);

    return (
        <ProductCardStyles>
            <Link className="card-image" to={content.search_name}>
                <img src={content.image.url} alt={content.name}/>
            </Link>
            <div className="card-text">
                <Link to={content.search_name}>{content.name}</Link>
                <p className="highlight">
                    {content.price.toFixed(2)}USD
                </p>
            </div>
            <Button
                variant={buttonVariant}
                size="small"
                label="Order Now"
                onClick={clickHandler}
            />
        </ProductCardStyles>
    );
};

export default ProductCard;