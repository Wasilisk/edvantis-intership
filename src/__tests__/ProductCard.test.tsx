import {render} from "./utils/store-wrapper";
import {fireEvent} from "@testing-library/react";
import ProductCard from "../components/ProductCard";
import {unstable_HistoryRouter as HistoryRouter} from "react-router-dom";
import {mockProductCard} from "./__mock__/mock-product-card";
import {createMemoryHistory} from "history";

describe('ProductCard component', () => {
    it('should render component', () => {
        const history = createMemoryHistory({initialEntries: ['/']});
        const {getByText, asFragment, getByAltText} = render(
            <HistoryRouter history={history}>
                <ProductCard content={mockProductCard} buttonVariant="filled"/>
            </HistoryRouter>
        );

        const cardImage = getByAltText(mockProductCard.name);
        expect(cardImage).toHaveAttribute('src', mockProductCard.image.url);
        fireEvent.click(cardImage);
        expect(history.location.pathname).toBe(`/${mockProductCard.search_name}`);

        const productName = getByText(mockProductCard.name);
        expect(productName).toBeDefined();

        const productPrice = getByText(`${mockProductCard.price.toFixed(2)}USD`);
        expect(productPrice).toBeDefined();

        expect(asFragment()).toMatchSnapshot();
    })
})