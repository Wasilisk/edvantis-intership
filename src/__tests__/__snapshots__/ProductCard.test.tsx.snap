// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductCard component should display out of stock block 1`] = `
<DocumentFragment>
  <div
    class="sc-bczRLJ eZuxEp"
  >
    <a
      class="card-image"
      href="/lord-armchair-miedel-home"
    >
      <img
        alt="Lord armchair, Miedel Home"
        src="https://e-c.storage.googleapis.com/res/5f731c835a5a2667cb27290e/480"
      />
      <div
        class="out-of-stock"
      >
        Out of stock
      </div>
    </a>
    <div
      class="card-text"
    >
      <a
        href="/lord-armchair-miedel-home"
      >
        Lord armchair, Miedel Home
      </a>
      <p
        class="highlight"
      >
        120.50USD
      </p>
    </div>
    <button
      class="sc-gsnTZi bxYKFi"
      disabled=""
    >
      Order Now
    </button>
  </div>
</DocumentFragment>
`;

exports[`ProductCard component should render component 1`] = `
<DocumentFragment>
  <div
    class="sc-bczRLJ eZuxEp"
  >
    <a
      class="card-image"
      href="/lord-armchair-miedel-home"
    >
      <img
        alt="Lord armchair, Miedel Home"
        src="https://e-c.storage.googleapis.com/res/5f731c835a5a2667cb27290e/480"
      />
    </a>
    <div
      class="card-text"
    >
      <a
        href="/lord-armchair-miedel-home"
      >
        Lord armchair, Miedel Home
      </a>
      <p
        class="highlight"
      >
        120.50USD
      </p>
    </div>
    <button
      class="sc-gsnTZi bxYKFi"
    >
      Order Now
    </button>
  </div>
</DocumentFragment>
`;
